cdc/sink/mysql.go

```diff
@@ -36,37 +37,168 @@ import (
-	db         *sql.DB
-	infoGetter TableInfoGetter
-	ddlOnly    bool
+	db               *sql.DB
+	globalResolvedTs uint64
+	sinkResolvedTs   uint64
+	checkpointTs     uint64
+
+	unresolvedRowsMu sync.Mutex
+	unresolvedRows   map[string][]*model.RowChangedEvent
 }
 
 func (s *mysqlSink) EmitResolvedEvent(ctx context.Context, ts uint64) error {
-	panic("implement me")
+	atomic.StoreUint64(&s.globalResolvedTs, ts)
+	return nil
 }
 
 func (s *mysqlSink) EmitCheckpointEvent(ctx context.Context, ts uint64) error {
-	panic("implement me")
+	return nil
+}
+
+func (s *mysqlSink) EmitRowChangedEvent(ctx context.Context, rows ...*model.RowChangedEvent) error {
+	var resolvedTs uint64
+	s.unresolvedRowsMu.Lock()
+	defer s.unresolvedRowsMu.Unlock()
+	for _, row := range rows {
+		if row.Resolved {
+			resolvedTs = row.Ts
+			continue
+		}
+		key := util.QuoteSchema(row.Schema, row.Table)
+		s.unresolvedRows[key] = append(s.unresolvedRows[key], row)
+	}
+	atomic.StoreUint64(&s.sinkResolvedTs, resolvedTs)
+	return nil
 }
 
-func (s *mysqlSink) EmitRowChangedEvent(ctx context.Context, txn ...*model.RowChangedEvent) error {
-	panic("implement me")
+func (s *mysqlSink) EmitDDLEvent(ctx context.Context, ddl *model.DDLEvent) error {
+	err := s.execDDLWithMaxRetries(ctx, ddl, 5)
+	return errors.Trace(err)
 }
 
-func (s *mysqlSink) EmitDDLEvent(ctx context.Context, txn *model.DDLEvent) error {
-	panic("implement me")
+func (s *mysqlSink) execDDLWithMaxRetries(ctx context.Context, ddl *model.DDLEvent, maxRetries uint64) error {
+	return retry.Run(func() error {
+		err := s.execDDL(ctx, ddl)
+		if isIgnorableDDLError(err) {
+			return nil
+		}
+		return err
+	}, maxRetries)
+}
+
+func (s *mysqlSink) execDDL(ctx context.Context, ddl *model.DDLEvent) error {
+	shouldSwitchDB := len(ddl.Schema) > 0 && ddl.Type != timodel.ActionCreateSchema
+
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if shouldSwitchDB {
+		_, err = tx.ExecContext(ctx, "USE "+util.QuoteName(ddl.Schema)+";")
+		if err != nil {
+			if rbErr := tx.Rollback(); rbErr != nil {
+				log.Error("Failed to rollback", zap.Error(err))
+			}
+			return errors.Trace(err)
+		}
+	}
+
+	if _, err = tx.ExecContext(ctx, ddl.Query); err != nil {
+		if rbErr := tx.Rollback(); rbErr != nil {
+			log.Error("Failed to rollback", zap.String("sql", ddl.Query), zap.Error(err))
+		}
+		return errors.Trace(err)
+	}
+
+	if err = tx.Commit(); err != nil {
+		return errors.Trace(err)
+	}
+
+	log.Info("Exec DDL succeeded", zap.String("sql", ddl.Query))
+	return nil
 }
 
 func (s *mysqlSink) CheckpointTs() uint64 {
-	panic("implement me")
+	return atomic.LoadUint64(&s.checkpointTs)
+}
+
+func (s *mysqlSink) Run(ctx context.Context) error {
+	execRows := func(rowsMap map[string][]*model.RowChangedEvent) error {
+		groups := make([][]*model.RowChangedEvent, 0, len(rowsMap))
+		for _, rows := range rowsMap {
+			groups = append(groups, rows)
+		}
+		return s.concurrentExec(ctx, groups)
```

Comment:
we don't need to convert the `map[string][]*model.RowChangedEvent` to `[][]*model.RowChangedEvent`, we can change the parameter of `concurrentExec` and pass the map to it directly